File: src/ai/builtin.ts
// src/ai/builtin.ts — KataGo 없을 때 쓰는 내장 휴리스틱 AI
import { groupAndLibs, idx, inBounds, legalMoves, tryPlay } from '../engine/board'
import type { GameState, Point, Stone } from '../engine/types'
import { getRank, type RankId } from './ranks'

export interface RankedMove {
  point: Point
  score: number
  label: string
}

/** 이 점이 color의 '진짜 눈'인지 — 상하좌우 전부 아군, 대각 적 점유가 기준 미만 */
function isTrueEye(board: Stone[], size: number, x: number, y: number, color: Stone): boolean {
  if (board[idx(size, x, y)] !== 0) return false
  const orth: Array<[number, number]> = [
    [x - 1, y],
    [x + 1, y],
    [x, y - 1],
    [x, y + 1],
  ]
  for (const [nx, ny] of orth) {
    if (!inBounds(size, nx, ny)) continue
    if (board[idx(size, nx, ny)] !== color) return false
  }
  const diag: Array<[number, number]> = [
    [x - 1, y - 1],
    [x + 1, y - 1],
    [x - 1, y + 1],
    [x + 1, y + 1],
  ]
  let enemyDiag = 0
  let offBoard = 0
  for (const [nx, ny] of diag) {
    if (!inBounds(size, nx, ny)) {
      offBoard += 1
      continue
    }
    const v = board[idx(size, nx, ny)]
    if (v !== 0 && v !== color) enemyDiag += 1
  }
  // 변·귀는 대각 1개만 적이어도 가짜 눈, 중앙은 2개부터
  const limit = offBoard > 0 ? 1 : 2
  return enemyDiag < limit
}

/**
 * 한 수 평가 (합법수 전수 재탐색 없이 가벼운 지표만):
 * 따냄 / 단수 걸기 / 단수 도망 / 자충 회피 / 초반 3·4선 선호
 */
function evaluateMove(state: GameState, p: Point, noise: number): number {
  const { size, board, toPlay } = state
  const opp: Stone = toPlay === 1 ? 2 : 1

  // 자기 진짜 눈 메우기는 절대 금지 (종반 자멸 방지)
  if (isTrueEye(board, size, p.x, p.y, toPlay)) return -Infinity

  const r = tryPlay(state, p.x, p.y)
  if (!r.ok) return -Infinity

  let score = r.move.captured.length * 40

  const after = r.state.board

  // 착수 후 내 그룹 활로 — 자충(스스로 단수) 강한 감점
  const own = groupAndLibs(after, size, p.x, p.y)
  if (own.libs.length === 1) score -= 30
  else if (own.libs.length === 2) score -= 4

  // 착수 전 인접 아군이 단수였는데 이 수로 살아나면 큰 가점 (단수 도망/이음)
  const orth: Array<[number, number]> = [
    [p.x - 1, p.y],
    [p.x + 1, p.y],
    [p.x, p.y - 1],
    [p.x, p.y + 1],
  ]
  let rescued = false
  const seenOppGroup = new Set<number>()
  let atariThreats = 0
  let pressure = 0
  let ownContact = 0
  for (const [nx, ny] of orth) {
    if (!inBounds(size, nx, ny)) continue
    const before = board[idx(size, nx, ny)]
    if (before === toPlay) {
      ownContact += 1
      if (!rescued) {
        const g = groupAndLibs(board, size, nx, ny)
        if (g.libs.length === 1 && own.libs.length >= 2) rescued = true
      }
    } else if (before === opp) {
      const rootAfter = after[idx(size, nx, ny)]
      if (rootAfter !== opp) continue // 이미 따낸 돌
      const key = idx(size, nx, ny)
      if (seenOppGroup.has(key)) continue
      const g = groupAndLibs(after, size, nx, ny)
      for (const s of g.stones) seenOppGroup.add(idx(size, s.x, s.y))
      if (g.libs.length === 1) atariThreats += 1
      else if (g.libs.length === 2) pressure += 1
    }
  }
  if (rescued) score += 34
  score += atariThreats * 14 + pressure * 4 + ownContact * 1

  // 초반 위치 감각: 3·4선(귀·변) 선호, 1·2선·과한 중앙 지양
  const early = state.history.length < size * 2
  const line = Math.min(p.x, p.y, size - 1 - p.x, size - 1 - p.y) + 1
  if (early) {
    if (line === 3 || line === 4) score += 6
    else if (line === 2) score += 1
    else if (line === 1) score -= 5
    else score += 2
  } else {
    const center = (size - 1) / 2
    const dist = Math.abs(p.x - center) + Math.abs(p.y - center)
    score += (size - dist) * 0.3
    if (line === 1) score -= 2
  }

  return score + (noise > 0 ? Math.random() * noise : 0)
}

/** 후보 목록 — 자기 눈 메우기 제외 */
function candidateMoves(state: GameState): Point[] {
  return legalMoves(state).filter(
    (p) => !isTrueEye(state.board, state.size, p.x, p.y, state.toPlay),
  )
}

/**
 * 내장 AI 착수 선택. null = 패스 권고.
 * - 의미 있는 수가 없고 상대가 패스했으면 패스해 대국을 끝냄
 */
export function pickBuiltinMove(state: GameState, rankId: RankId): Point | null {
  const moves = candidateMoves(state)
  if (moves.length === 0) return null

  const rank = getRank(rankId)
  // 레벨이 낮을수록 평가에 노이즈를 더해 실수 유도
  const noise = Math.max(0, (10 - rank.level) * 1.2)
  let best: Point | null = null
  let bestScore = -Infinity
  for (const m of moves) {
    const s = evaluateMove(state, m, noise)
    if (s > bestScore) {
      bestScore = s
      best = m
    }
  }
  if (best === null) return null

  // 상대가 패스했고 따냄·구출 같은 급한 수가 없으면 레벨 무관 같이 패스 (계가로 종료)
  if (state.consecutivePasses >= 1 && bestScore < 10) return null

  if (Math.random() < rank.randomness) {
    return moves[Math.floor(Math.random() * moves.length)]
  }
  return best
}

/** 힌트/분석: 상위 n수 + 숫자 라벨 (랜덤 없음) */
export function pickBuiltinTopMoves(state: GameState, n = 3): RankedMove[] {
  const moves = candidateMoves(state)
  if (moves.length === 0) return []
  const ranked = moves
    .map((point) => ({ point, score: evaluateMove(state, point, 0) }))
    .filter((m) => Number.isFinite(m.score))
    .sort((a, b) => b.score - a.score)
    .slice(0, Math.max(1, n))
  const best = ranked[0]?.score || 1
  return ranked.map((m, i) => {
    const pct = Math.max(1, Math.round((m.score / Math.max(best, 1)) * 100))
    return {
      point: m.point,
      score: m.score,
      label: `${i + 1}·${pct}`,
    }
  })
}

File: app/lib/domain/ai/builtin.dart
// lib/domain/ai/builtin.dart — KataGo 없이 두는 휴리스틱 AI
//
// TS 판(src/ai/builtin.ts)에서 두 가지를 고쳐 옮겼다.
//
// 1. scoreMove 와 scoreMoveDeterministic 이 흔들림 항 하나만 다른
//    복붙 쌍이었다. 한쪽만 고치면 힌트와 실제 착수가 갈라진다 → 한 함수로.
// 2. Math.random() 직접 호출이라 "1000판 반칙 0" 을 재현 가능하게
//    검증할 수 없었다 → Random 을 주입받는다.

import 'dart:math' as math;

import '../engine/board.dart';
import '../engine/types.dart';
import 'ranks.dart';

class RankedMove {
  const RankedMove({required this.point, required this.score, required this.rank, required this.percent});
  final Point point;
  final double score;

  /// 1부터 — 화면·낭독의 "1순위"
  final int rank;

  /// 최고 수 대비 백분율 (1~100)
  final int percent;
}

/// 한 수의 가치. [jitter] 가 0 이면 완전 결정적이다 (힌트·테스트용).
double _scoreMove(GameState state, Point p, double jitter, math.Random rng) {
  final PlayResult r = tryPlay(state, p.x, p.y);
  if (r is! PlayOk) return double.negativeInfinity;

  double score = r.move.captured.length * 50.0;

  // 중앙 선호 — 초심자에게 자연스러운 모양이 나온다
  final double center = (state.size.lines - 1) / 2;
  final double dist = (p.x - center).abs() + (p.y - center).abs();
  score += (state.size.lines - dist) * 0.5;

  // 1선은 약간 피한다
  final int last = state.size.lines - 1;
  if (p.x == 0 || p.y == 0 || p.x == last || p.y == last) score -= 2;

  // 상대에게 따냄 기회를 많이 주는 수는 감점.
  // capturePoints 가 전수 시뮬레이션과 같은 답을 O(판 크기) 로 준다.
  final GameState after = r.state;
  if (after.toPlay == state.toPlay.opponent) {
    if (capturePoints(after).length > 2) score -= 3;
  }

  return jitter == 0 ? score : score + rng.nextDouble() * jitter;
}

/// 내장 AI 의 한 수. 둘 곳이 없으면 null (호출자가 패스로 처리).
Point? pickBuiltinMove(GameState state, String rankId, {math.Random? rng}) {
  final math.Random r = rng ?? math.Random();
  final List<Point> moves = legalMoves(state);
  if (moves.isEmpty) return null;

  final RankOption rank = getRank(rankId);
  // 저레벨은 여기서 끝난다 — 탐색을 아예 안 하므로 19줄에서도 즉시 응답
  if (r.nextDouble() < rank.randomness) {
    return moves[r.nextInt(moves.length)];
  }

  Point best = moves.first;
  double bestScore = double.negativeInfinity;
  for (final Point m in moves) {
    final double s = _scoreMove(state, m, 2, r);
    if (s > bestScore) {
      bestScore = s;
      best = m;
    }
  }
  return best;
}

/// 힌트·분석용 상위 n수. 무작위 항이 없어 같은 판이면 항상 같은 답이 나온다.
List<RankedMove> pickBuiltinTopMoves(GameState state, {int n = 3}) {
  final List<Point> moves = legalMoves(state);
  if (moves.isEmpty) return const <RankedMove>[];

  final math.Random unused = math.Random(0); // jitter 0 이라 쓰이지 않는다
  final List<(Point, double)> scored = <(Point, double)>[];
  for (final Point p in moves) {
    final double s = _scoreMove(state, p, 0, unused);
    if (s.isFinite) scored.add((p, s));
  }
  if (scored.isEmpty) return const <RankedMove>[];

  scored.sort((( Point, double) a, (Point, double) b) => b.$2.compareTo(a.$2));
  final List<(Point, double)> top = scored.take(math.max(1, n)).toList();
  final double best = math.max(top.first.$2, 1);

  return <RankedMove>[
    for (int i = 0; i < top.length; i++)
      RankedMove(
        point: top[i].$1,
        score: top[i].$2,
        rank: i + 1,
        percent: math.max(1, (top[i].$2 / best * 100).round()),
      ),
  ];
}
